Persist Bindings annotations alongside pod Host assignment

jdef opened the pull request:
- resolves #4103

as per @bgrant0607:

> Status must be completely reconstructable based on observations. The proposed Pod.Status.Annotations would not be. Furthermore, I don't think we should introduce another annotations field to Pod. I'd prefer to add a podAnnotations field to Binding, and then merge them into the pod. As for clearing the annotations, I'd prefer to simply fail a pod that was rejected by a node so as not to have to deal with that.

In this PR, Bindings.Annotations are merged with those of the Pod, overwriting the values of any pre-existing keys. If the binding fails, the pod is flagged as failed.

Also added error checking in Binding creation, returning a more appropriate Status object on errors.

cc @smarterclayton @erictune

Commit:
resolves #4103
clarify resource conflict status, rebase to master
remove ResourceConflict, replace usage with Conflict

jdef commented:
xref https://github.com/mesosphere/kubernetes-mesos/issues/73

jdef commented:
FWIW I've already signed the CLA

jdef commented:
xref https://github.com/mesosphere/kubernetes-mesos/issues/162

smarterclayton commented:
I can do the review on this

jdef commented:
I applied this patch to the k8s code that I'm running in kubernetes-mesos and the annotations are still not showing up in the pod: they're missing in the pod json returned by the apiserver and are also not present in etcd. Digging in...

bgrant0607 commented on pkg/registry/pod/etcd/etcd.go:
```diff
@@ -191,10 +205,23 @@ func (r *BindingREST) assignPod(ctx api.Context, podID string, machine string) e
-		// can't really be helped, since there's not really a way to do atomic
-		// multi-object changes in etcd.
-		if _, err2 := r.setPodHostTo(ctx, podID, machine, ""); err2 != nil {
+		glog.V(1).Infof("failing pod %v because constraint checks failed", podID)
```
Please don't make this v1. It will happen too frequently. I suggest v4.

bgrant0607 commented:
Gratefully punting review to @smarterclayton, who is also touching /bindings right now.

This behavior needs to be documented, ideally somewhere that would appear in the swagger (e.g., in the description on the ObjectMeta field for Binding), as well as in a .md doc. Note that descriptions on fields for v1beta3 are in flight.

jdef commented:
related work in #5054

smarterclayton commented:
Sorry for the delay, will get to this very soon.

smarterclayton commented on pkg/registry/pod/etcd/etcd.go:
```diff
@@ -191,10 +205,23 @@ func (r *BindingREST) assignPod(ctx api.Context, podID string, machine string) e
+			err2 = r.store.Helper.AtomicUpdate(podKey, &api.Pod{}, false, func(obj runtime.Object) (runtime.Object, error) {
+				if pod, ok := obj.(*api.Pod); !ok {
+					return nil, fmt.Errorf("unexpected object: %#v", obj)
+				} else if pod.Status.Host != machine {
```
Can you remove the else's here?  Since you're returning they're not necessary and it fits our style more (if the if block returns, never use else).

smarterclayton commented:
Just my comment and Brian's I think.  Sorry again for the delay.

bgrant0607 commented:
Fails gofmt:
```
!! gofmt needs to be run on the following files: 
./pkg/registry/pod/etcd/etcd.go
```

bgrant0607 commented on pkg/registry/pod/etcd/etcd.go:
```diff
@@ -191,10 +205,24 @@ func (r *BindingREST) assignPod(ctx api.Context, podID string, machine string) e
+				} 
+				pod.Status.Host = "" //TODO(jdef) get rid of this once BoundPods goes away
+				pod.Status.Phase = api.PodFailed
+				pod.Status.Message = fmt.Sprintf("Assignment to host %v failed", machine)
```
Sorry, I was too terse last time. Please also populate Reason with something like `HostAssignmentFailed`. The more detailed message is useful, but the concise form (in particular, without the hostname) is useful for stats collection.

jdef commented:
@bgrant0607 It looks like reason codes are pretty specific and should closely track HTTP codes. So I changed the pod.status.message to what you suggested, and set the api.status.reason to Conflict. The api.status.message will include a more detailed description of what went wrong.

Sound ok?

smarterclayton commented on pkg/registry/pod/etcd/etcd.go:
```diff
@@ -142,15 +142,24 @@ func (r *BindingREST) New() runtime.Object {
+	err = r.assignPod(ctx, binding.PodID, binding.Host, binding.Annotations)
 	err = etcderr.InterpretCreateError(err, "binding", "")
-	out = &api.Status{Status: api.StatusSuccess}
+	if err != nil {
```
Hrm, this is somewhat wrong, because InterpretCreateError may already create an appropriate api.Status.  If assignPod needs to return special errors, that should happen in that method (using errors.NewConflict)

jdef commented:
@smarterclayton just re-read your comment after I pushed the "clean up error handling" change. PTAL and let me know if you still think I should create NewConflict errors in assignPod.

smarterclayton commented on pkg/registry/pod/etcd/etcd.go:
```diff
@@ -146,15 +147,20 @@ func (r *BindingREST) New() runtime.Object {
-	err = etcderr.InterpretCreateError(err, "binding", "")
+	err = r.assignPod(ctx, binding.PodID, binding.Host, binding.Annotations)
+	if err = etcderr.InterpretCreateError(err, "binding", ""); err != nil {
+		if _, ok := err.(*errors.StatusError); !ok {
```
This is fine for now.  Bindings internal storage will be cleanup up over successive iterations so we don't need the perfect solution now.

smarterclayton commented:
This looks good to me - can you go ahead and squash?

jdef commented:
squashed

On Mon, Mar 9, 2015 at 11:30 AM, Clayton Coleman <notifications@github.com>
wrote:

> This looks good to me - can you go ahead and squash?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5005#issuecomment-77875688>
> .
>



-- 
James DeFelice
585.241.9488 (voice)
650.649.6071 (fax)

smarterclayton commented:
Will merge after tests pass again.

smarterclayton commented:
Hrm, seeing a transient failure that looks related to this pull (not sure if it's coincidence or enemy action):

```
--- FAIL: TestEtcdCreateBindingNoPod (0.00s)
	<autogenerated>:29: returning /registry/pods/default/foo: &etcd.Response{Action:"", Node:(*etcd.Node)(nil), PrevNode:(*etcd.Node)(nil), EtcdIndex:0x0, RaftIndex:0x0, RaftTerm:0x0} &etcd.EtcdError{ErrorCode:100, Message:"", Cause:"", Index:0x0}
	etcd_test.go:873: Unexpected error returned: &errors.StatusError{ErrStatus:api.Status{TypeMeta:api.TypeMeta{Kind:"", APIVersion:""}, ListMeta:api.ListMeta{SelfLink:"", ResourceVersion:""}, Status:"Failure", Message:"binding \"\" cannot be updated: 100:  () [0]", Reason:"Conflict", Details:(*api.StatusDetails)(0xc20817eec0), Code:409}}
FAIL
coverage: 87.1% of statements
```

jdef commented:
Funny, I think it's related to another PR of mine that was just merged. I'll take a look

jdef commented:
@smarterclayton I'm not sure what's holding up the coveralls test. From what I can see the tests pass.

smarterclayton commented:
K, will look in a bit


> On Mar 10, 2015, at 3:58 PM, James DeFelice <notifications@github.com> wrote:
> 
> @smarterclayton I'm not sure what's holding up the coveralls test. From what I can see the tests pass.
> 
> —
> Reply to this email directly or view it on GitHub.
>

bgrant0607 commented on pkg/registry/pod/etcd/etcd.go:
```diff
@@ -203,14 +218,28 @@ func (r *BindingREST) assignPod(ctx api.Context, podID string, machine string) e
+				}
+				pod.Status.Host = "" //TODO(jdef) get rid of this once BoundPods goes away
+				pod.Status.Phase = api.PodFailed
+				pod.Status.Message = "HostAssignmentFailed"
```
Pod status message should be more informative, such as including the host. Search the code for examples.

bgrant0607 commented on pkg/registry/pod/etcd/etcd.go:
```diff
@@ -183,14 +189,23 @@ func (r *BindingREST) setPodHostTo(ctx api.Context, podID, oldMachine, machine s
+		err = etcderr.InterpretGetError(err, "pod", podID)
+		err = etcderr.InterpretUpdateError(err, "pod", podID)
+		if _, ok := err.(*errors.StatusError); !ok && err != nil {
+			err = errors.NewConflict("binding", "", err)
```
I'm concerned about returning conflict for constraint failures. Clients may retry assuming it's a resourceVersion mismatch.

bgrant0607 replied:
We should create a new StatusReason -- we already support 2 different reasons for 409. How about "ResourceConflict"? It's a bit unfortunate that "resource" is ambiguous due to usage in the REST context, but I can envision conflicts on other underlying infrastructure resources (e.g., networking) in the future.

bgrant0607 commented:
Don't worry about coveralls. It appears to be wedged.

jdef commented:
Thanks for the feedback @bgrant0607. I've pushed some changes to address your concerns.

jdef commented:
PTAL @bgrant0607 @smarterclayton

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1025,6 +1025,13 @@ const (
 	StatusReasonConflict StatusReason = "Conflict"
 
+	// StatusReasonResourceConflict means the requested update operation cannot be completed
+	// due to a resource conflict in the operation. The client may need to alter the request.
```
How about "due to conflicting usage of underlying system/infrastructure resources".

jdef replied:
done

On Wed, Mar 11, 2015 at 6:29 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/api/types.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5005#discussion_r26262653>
> :
>
> > @@ -1025,6 +1025,13 @@ const (
> >  	// Status code 409
> >  	StatusReasonConflict StatusReason = "Conflict"
> >
> > +	// StatusReasonResourceConflict means the requested update operation cannot be completed
> > +	// due to a resource conflict in the operation. The client may need to alter the request.
>
> How about "due to conflicting usage of underlying system/infrastructure
> resources".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5005/files#r26262653>
> .
>



-- 
James DeFelice
585.241.9488 (voice)
650.649.6071 (fax)

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1025,6 +1025,13 @@ const (
 
+	// StatusReasonResourceConflict means the requested update operation cannot be completed
+	// due to a resource conflict in the operation. The client may need to alter the request.
+	// Each resource may define custom details that indicate the nature of the
```
"Each API resource..."

This is confusing because we're using the term "resource" with 2 different meanings, but I couldn't think of a better word.

bgrant0607 commented on pkg/registry/pod/etcd/etcd.go:
```diff
@@ -203,12 +215,31 @@ func (r *BindingREST) assignPod(ctx api.Context, podID string, machine string) e
+		glog.V(4).Infof("failing pod %v because constraint checks failed", podID)
+		podKey, err2 := r.store.KeyFunc(ctx, podID)
+		if err2 == nil {
+			err2 = r.store.Helper.AtomicUpdate(podKey, &api.Pod{}, false, func(obj runtime.Object) (runtime.Object, error) {
```
Why can't you use setPodHostAndAnnotations?

jdef replied:
I'll consolidate. I thought it would be uglier adding an extra `failed` or `rollback` parameter to `setPodHostAndAnnotations`, but it would keep the pod host & status update code all in one place -- which I think is what you're aiming for, yes?

bgrant0607 replied:
Yes, thanks.

bgrant0607 commented on pkg/registry/pod/etcd/etcd.go:
```diff
@@ -176,6 +176,12 @@ func (r *BindingREST) setPodHostTo(ctx api.Context, podID, oldMachine, machine s
 			return nil, fmt.Errorf("pod %v is already assigned to host %v", pod.Name, pod.Status.Host)
 		}
 		pod.Status.Host = machine
```
You'll need to rebase this FYI.

jdef replied:
done

On Wed, Mar 11, 2015 at 6:33 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/registry/pod/etcd/etcd.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5005#discussion_r26263030>
> :
>
> > @@ -176,6 +176,12 @@ func (r *BindingREST) setPodHostTo(ctx api.Context, podID, oldMachine, machine s
> >  			return nil, fmt.Errorf("pod %v is already assigned to host %v", pod.Name, pod.Status.Host)
> >  		}
> >  		pod.Status.Host = machine
>
> You'll need to rebase this FYI.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5005/files#r26263030>
> .
>



-- 
James DeFelice
585.241.9488 (voice)
650.649.6071 (fax)

jdef commented:
addressed your concerns @bgrant0607 PTAL

bgrant0607 commented:
Code LGTM.

Are you able to run the e2e test? The change to write Spec.Host in setPodHostTo broke it, due to somewhat fragile assumptions.

We just added documentation of status codes to docs/api-conventions.md. Could you please add documentation on ResourceConflict there?

bgrant0607 commented:
Alternatively, we could remove ResourceConflict and change back to Conflict. Host mismatch or etcd collision are the only reasons for failure now, since we removed hostport conflict detection.

We'll likely want to add back some best-effort constraint checking at some point. See:
https://github.com/GoogleCloudPlatform/kubernetes/issues/5207#issuecomment-78613182

But we don't have it now.

I'll leave it up to you.

cc @jszczepkowski

jdef commented:
Removed ResourceConflict. I don't want to muddy the water with additional
constraint checking here.

On Thu, Mar 12, 2015 at 9:43 PM, Brian Grant <notifications@github.com>
wrote:

> Alternatively, we could remove ResourceConflict and change back to
> Conflict. Host mismatch or etcd collision are the only reasons for failure
> now, since we removed hostport conflict detection.
>
> We'll likely want to add back some best-effort constraint checking at some
> point. See:
> #5207 (comment)
> <https://github.com/GoogleCloudPlatform/kubernetes/issues/5207#issuecomment-78613182>
>
> But we don't have it now.
>
> I'll leave it up to you.
>
> cc @jszczepkowski <https://github.com/jszczepkowski>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5005#issuecomment-78739798>
> .
>



-- 
James DeFelice
585.241.9488 (voice)
650.649.6071 (fax)

jdef commented:
@bgrant0607 any final thoughts?

jdef commented:
PTAL @bgrant0607 @smarterclayton

bgrant0607 commented:
LGTM
